index.js

```diff
@@ -143,42 +145,66 @@ module.exports = {
         return cb(new Error('addMembers only supports bendybutt-v1 feed IDs'))
       }
 
-      get(groupId, (err, { writeKey, root }) => {
+      get(groupId, (err, { root }) => {
         // prettier-ignore
         if (err) return cb(clarify(err, `Failed to get group details when adding members`))
 
         getRootFeedIdFromMsgId(root, (err, rootAuthorId) => {
           // prettier-ignore
           if (err) return cb(clarify(err, "couldn't get root id of author of root msg"))
 
-          const content = {
-            type: 'group/add-member',
-            version: 'v2',
-            groupKey: writeKey.key.toString('base64'),
-            root,
-            creator: rootAuthorId,
-            recps: [groupId, ...feedIds],
-          }
+          getTipEpochs(groupId, (err, tipEpochs) => {
+            // prettier-ignore
+            if (err) return cb(clarify(err, "Couldn't get tip epochs when adding members"))
 
-          if (opts.text) content.text = opts.text
+            const getFeed = opts?._feedKeys
+              ? (cb) => cb(null, { keys: opts._feedKeys })
+              : findOrCreateAdditionsFeed
 
-          const getFeed = opts?._feedKeys
-            ? (cb) => cb(null, { keys: opts._feedKeys })
-            : findOrCreateAdditionsFeed
+            getFeed((err, additionsFeed) => {
+              // prettier-ignore
+              if (err) return cb(clarify(err, 'Failed to find or create additions feed when adding members'))
 
-          getFeed((err, additionsFeed) => {
-            // prettier-ignore
-            if (err) return cb(clarify(err, 'Failed to find or create additions feed when adding members'))
+              const options = {
+                isValid: isAddMember,
+                tangles: ['members'],
+                feedKeys: additionsFeed.keys,
+              }
+              pull(
+                pull.values(tipEpochs),
+                pull.asyncMap((tipEpoch, cb) => {
+                  getPredecessorEpochs(
+                    groupId,
+                    tipEpoch.id,
+                    (err, predecessors) => {
+                      // prettier-ignore
+                      if (err) return cb(clarify(err, "Couldn't get predecessor epochs when adding members"))
+
+                      const oldSecrets = predecessors.map((pred) =>
+                        pred.secret.toString('base64')
+                      )
+                      const content = {
+                        type: 'group/add-member',
+                        version: 'v2',
+                        secret: tipEpoch.secret.toString('base64'),
+                        oldSecrets: opts.oldSecrets ? oldSecrets : undefined,
```

Comment:
:fire: I would rather that this is ALWAYS an array, even if empty.
This is something a schema check would catch

Reply:
oh... `undefined` fields will get dropped on stringification.
But we already know this should not be empty in the case where we're in a second epoch...?

Reply:
there's 3 cases sort of
1. addition to first epoch: no old secrets yet
2. addition to later epoch: old secrets
3. re-addition after epoch rotation: old secrets not needed to be sent

and they all have the same validator atm. which i think is sort of fine?

Reply:
yeah same validator is fine. In that case I think we should just always pass an Array - this was you can always call into the field and iterate over it and no-one ever falls in a hole and we don't need to put `?.` everywhere

Reply:
:fire: I think this should always be an `Array`

Reply:
Bah, have put this in an issue to keep this PR moving:
https://github.com/ssbc/ssb-tribes2/issues/118